src/Foundatio.AWS/Queues/SQSQueue.cs

```diff
@@ -0,0 +1,269 @@
+
+namespace Foundatio.AWS.Queues {
+    public class SQSQueue<T> : QueueBase<T> where T : class {
+
+        private readonly AsyncLock _lock = new AsyncLock();
+
+        private readonly Lazy<AmazonSQSClient> _client;
+        private readonly TimeSpan _workItemTimeout;
+        private readonly TimeSpan _readQueueTimeout;
+        private readonly AWSCredentials _credentials;
+        private readonly RegionEndpoint _regionEndpoint;
+
+        private string _queueUrl;
+
+        private long _enqueuedCount;
+        private long _dequeuedCount;
+        private long _completedCount;
+        private long _abandonedCount;
+        private long _workerErrorCount;
+
+
+        public SQSQueue(
+            string queueName,
+            AWSCredentials credentials = null,
+            RegionEndpoint regionEndpoint = null,
+            TimeSpan? workItemTimeout = null,
+            TimeSpan? readQueueTimeout = null,
+            ISerializer serializer = null,
+            IEnumerable<IQueueBehavior<T>> behaviors = null,
+            ILoggerFactory loggerFactory = null)
+            : base(serializer, behaviors, loggerFactory) {
+
+            _queueName = queueName ?? throw new ArgumentNullException(nameof(queueName));
+
+            _credentials = credentials ?? FallbackCredentialsFactory.GetCredentials();
+            _regionEndpoint = regionEndpoint ?? RegionEndpoint.USEast1;
+            _workItemTimeout = workItemTimeout ?? TimeSpan.FromMinutes(5);
+            _readQueueTimeout = readQueueTimeout ?? TimeSpan.FromSeconds(2);
+
+            _client = new Lazy<AmazonSQSClient>(() => new AmazonSQSClient(_credentials, _regionEndpoint));
+        }
+
+        protected override async Task EnsureQueueCreatedAsync(CancellationToken cancellationToken = new CancellationToken()) {
+            if (!String.IsNullOrEmpty(_queueUrl))
+                return;
+
+            using (await _lock.LockAsync(cancellationToken).ConfigureAwait(false)) {
+                if (!String.IsNullOrEmpty(_queueUrl))
+                    return;
+
+                // per sqs documentation, create will get existing or create new queue
+                var createRequest = new CreateQueueRequest { QueueName = _queueName };
+                var createResponse = await _client.Value.CreateQueueAsync(createRequest, cancellationToken).ConfigureAwait(false);
+
+                _queueUrl = createResponse.QueueUrl;
+            }
+        }
+
+        protected override async Task<string> EnqueueImplAsync(T data) {
+            if (!await OnEnqueuingAsync(data).ConfigureAwait(false))
+                return null;
+
+            Interlocked.Increment(ref _enqueuedCount);
+            var message = new SendMessageRequest {
+                QueueUrl = _queueUrl,
+                MessageBody = await _serializer.SerializeToStringAsync(data).ConfigureAwait(false),
+            };
+
+            var response = await _client.Value.SendMessageAsync(message).ConfigureAwait(false);
```

Comment:
Should we be retrying this? We retry in half of our current implementations (curious what your thoughts are) https://github.com/exceptionless/Foundatio/blob/master/src/Foundatio.Redis/Queues/RedisQueue.cs#L201